### ISSUE TITLE
RS Controllers cause lots of lag

### ISSUE THREAD

**Issue by kreezxil** [raoulvdberge/refinedstorage#2248]
#### Issue description:
On my server for World of Dragons, I've using Lag Goggles located and eliminate sources of lag. Consitently I'm finding that the RS controller is a major source of it.  Further I'm finding that If I set them to not work except when redstone signal is present they lag even more.

rs controller off
![image](https://user-images.githubusercontent.com/3880072/58372549-98a36880-7ee4-11e9-9728-cf2cb9e37650.png)

rs controller on
![image](https://user-images.githubusercontent.com/3880072/58372630-12882180-7ee6-11e9-81a8-daf00d5edf09.png)

if you need an .nps report let me know and I'll get one going.

**Comment by raoulvdberge** [comment]
I need one

**Comment by P3rf3ctXZer0** [comment]
Can I see your complete modlist? You can print a dirctory text file using this application. https://www.softpedia.com/get/System/File-Management/DirLister.shtml

**Comment by kreezxil** [comment]
http://paste.ubuntu.com/p/CVGpzrzqKH/

**Comment by raoulvdberge** [comment]
?

I need a NPS log

**Comment by kreezxil** [comment]
hold on please. getting around to it. zero wanted to see my modlist.

**Comment by P3rf3ctXZer0** [comment]
Nah its fine. @kreezxil I cannot help with this issue.

**Comment by raoulvdberge** [comment]
Ah, @P3rf3ctXZer0 deleted his comment so I couldn't see he was asking.

**Comment by kreezxil** [comment]
I have sampler installed, nps is coming. Here's an rs controller doing 33 ms/t as reported by lag goggles. I've found some as high as 1500 ms/t.

![image](https://user-images.githubusercontent.com/3880072/58508302-59318200-8159-11e9-916e-d8b20e9f296f.png)


[sampler-snapshot.zip](https://github.com/raoulvdberge/refinedstorage/files/3229464/sampler-snapshot.zip)

here it is again, but this time I told it to only work with redstone signal and put lever on it in the off state. It's dong 216 ms/t lag now.

![image](https://user-images.githubusercontent.com/3880072/58508565-eb398a80-8159-11e9-8178-0dc68e9e6064.png)

and another nps file

[sampler-snapshot.zip](https://github.com/raoulvdberge/refinedstorage/files/3229481/sampler-snapshot.zip)

**Comment by Darkere** [comment]
Laggogles does not work well with Tile Entities that interact with the event bus. It then calculates things completely wrong often ending up with numbers way larger than what is actually there.

I'm also not sure you are aware of how little 1000 μs/t is. Minecraft still runs fine without lag until 50000 μs/t. 

Your sampling shows this nicely:
![image](https://user-images.githubusercontent.com/4283717/58511964-35dbf680-819c-11e9-81dc-4c55ae07c4c0.png)

You can see the controller at 30ms which is about the same as your cloches and half of what the tech reborn pipes take that are on the server (In the 10s that the server ran during sampling). 

Start a crafting task that has more than 10000 steps and open the crafting monitor to see how much lag RS can produce. We know and are working on it.

**Comment by kreezxil** [comment]
For me tho, I have about 1000 players on my servers, 20 can be logged on at
any given time. During the weekends and holidays the servers are packed to
the gills. During the week not so much. 30ms/t adds up. that was the small
amount. Also you didn't so exactly how long to run the sampler for so I did
for about 5 minutes. I will run a deeper one over the weekend.

**Comment by Darkere** [comment]
The file that you uploaded only ran for 10 seconds so you might have grabbed the wrong one?

Ok let me rephrase how the performance plays out:

You ran sampler for roughly 10 Seconds.
-The server was Idle for 1.5 seconds so roughly 15% of the time.
-The server spend 4 seconds updating Entities, 1/3 of that on fossils and archeology
-Another 2 seconds were spent un-/loading chunks
-Another second went figuring out players and where entities can be spawned

Then we dig deep and find Refined Storage at around 0.14 seconds total. This is for all RS systems on the server combined. 

That means it spent 1% of its time calculating RS. If you are looking for better performance you are in the wrong place.

**Comment by kreezxil** [comment]
I'm here reporting what I felt is an issue to RS team. And to learn more
about NPS considering this team is telling me that lag goggles is basically
garbage.

**Comment by Darkere** [comment]
Laggoggles is by no means garbage. It's really nice to find many basic issues, many of these are not really avoidable for (inexperienced) mod authors and can be easily fixed by the player. 

But It does come with quite a few limitations and bugs. You need some idea of what you are looking at and some healthy skepticism of what it is reporting. 
In RS's case, for example, almost every calculation for the whole system (+ what comes from a bug in LG) is done inside the controller. That means that it's gonna be higher than many others that only show the performance of one block. 
Laggoggles cannot know that, hence the red color. 

In doubt (or when the lag comes from the event bus) run sampler or the much easier to read and share [spark](https://github.com/lucko/spark) (if you set the mappings in the top right..).

**Comment by kreezxil** [comment]
I use spark too.

**Comment by CraftalotReloaded** [comment]
Loading a system with about 20 million items takes like 80sec, that's how bad the performance is, just saying

**Comment by kreezxil** [comment]
I can hear it now.

**BUT WHY DO YOU NEED 20 MILLION ITEMS**?
